xds/internal/resolver/serviceconfig_action_test.go

```diff
@@ -0,0 +1,252 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package resolver
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	xdsclient "google.golang.org/grpc/xds/internal/client"
+)
+
+func TestNewActionsFromRoutes(t *testing.T) {
+	tests := []struct {
+		name   string
+		routes []*xdsclient.Route
+		want   map[string]action
+	}{
+		{
+			name: "temp",
+			routes: []*xdsclient.Route{
+				{Action: map[string]uint32{"B": 60, "A": 40}},
+				{Action: map[string]uint32{"A": 30, "B": 70}},
+				{Action: map[string]uint32{"B": 90, "C": 10}},
+			},
+			want: map[string]action{
+				"A40_B60_": {map[string]uint32{"A": 40, "B": 60}, "A_B_", ""},
+				"A30_B70_": {map[string]uint32{"A": 30, "B": 70}, "A_B_", ""},
+				"B90_C10_": {map[string]uint32{"B": 90, "C": 10}, "B_C_", ""},
+			},
+		},
+	}
+
+	cmpOpts := []cmp.Option{cmp.AllowUnexported(action{})}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := newActionsFromRoutes(tt.routes)
+			if !cmp.Equal(got, tt.want, cmpOpts...) {
+				t.Errorf("newActionsFromRoutes() got unexpected result, diff %v", cmp.Diff(got, tt.want, cmpOpts...))
+			}
+		})
+	}
+}
+
+func TestRemoveOrReuseName(t *testing.T) {
+	tests := []struct {
+		name          string
+		oldActions    map[string]action
+		oldNextIndex  map[string]int
+		newActions    map[string]action
+		wantActions   map[string]action
+		wantNextIndex map[string]int
```

Comment:
Testing the value of `wantNextIndex` seems to be testing implementation rather than behavior. And if we go with my other suggestion of using random numbers for the next index, we wouldn't have to test that at all.

Reply:
It to make sure that the indices are incrementing, so it won't generate the same assigned names.

Random number would make this, and the expected assigned names hard to test.